gef.py

```diff
@@ -2593,6 +2593,79 @@ def inner_f(*args, **kwargs):
     return wrapper
 
 
+def FakeExit(*args, **kwargs):
+    raise RuntimeWarning
+
+sys.exit = FakeExit
+
+def parse_arguments(required_arguments, optional_arguments):
+    """Argument parsing decorator."""
+
+    def int_wrapper(x): return int(x, 0)
+
+    def decorator(f):
+        def wrapper(*args, **kwargs):
+            parser = argparse.ArgumentParser(prog=args[0]._cmdline_, add_help=True)
+            for argname in required_arguments:
+                argvalue = required_arguments[argname]
+                argtype = type(argvalue)
+                if argtype == int:
+                    argtype = int_wrapper
+
+                if argname.startswith("-"):
```

Comment:
How can we have optional args in the `required_arguments` part?

Reply:
We shouldn't have this section.

Why would we support this:

`@parse_arguments({"--x": 0}, {})`

AND

`@parse_arguments({}, {"--x": 0})`

(which do the same thing)

and

`@parse_arguments({"x": 0}, {})`

but not

`@parse_arguments({}, {"x": 0})`

We should either:

1. Take 2 arguments. Everything in arg 1 is required, everything in arg2 is option
2. Take 1 argument. Everything that starts with `--` is optional, everything missing it is required.

Reply:
I understand your points better, so let me clarify: I made a lot of implementation choices on best effort towards what would be the most useful. I don't claim they are perfect.

The approach `parse_arguments( required_dict, optional_dict)` was made to easily allow creating mandatory arguments: a lot of commands that don't use `getopt` still do some checks over `argv`, having required positional arguments. This was meant to (in the future) wraps all those commands easily. Basic (non-existing) example would be replacing

```py
def do_invoke(self, argv):
   if len(argv) != 1: 
      err(...)
      return
   ...
```

To 

```py
@parse_arguments({"whatever", ""})
def do_invoke(self, argv, *args, **kwargs):
```

Also non-required arguments cannot be positional because I allowed `nargs='*'` which is useful (see `AssembleCommand` for an example). Hence the check for non-required args starting with `-`.

Regarding what you said about making the required dict in the format `{"name" : type }`, I don't like, I find it ugly and confusing. Also since I allowed the use of `nargs='*'` you **need** default values (think of the case you have 2 `nargs='*'`). But if @daniellimws feels the same way, I'll follow the majority.

> We should either:
> 
> 1. Take 2 arguments. Everything in arg 1 is required, everything in arg2 is option

Yes that's exactly what is there now, with the fact that arg1 also accepts `--args` that are marked as `required=true`.

> 2. Take 1 argument. Everything that starts with -- is optional, everything missing it is required.

No that's a simplistic approach: using `nargs='*'` prevents from having multiple position ` nargs='*'` arguments (I've tried, if you find a way, I'm open). But on the other hand, you may want to pass other options that you (command developer) want required. Hence the possibility to provide required `--foo`.

Think of the `AssembleCommand`: you wanna do `asm --arch=x86 nop; nop; nop;` I used positional argument with `nargs='*'` to collect all instructions. But say I want to force the user to provide an architecture, with your approach it wouldn't work.

Hope it helps understand.

Reply:
Yes, that helps a lot! Do we have any commands that are positional that require a `-` prefix? I think we could avoid that, and then simply put positional (and variadic) args in the first and optional in the second. Does that work?

Reply:
> Regarding what you said about making the required dict in the format {"name" : type }, I don't like, I find it ugly and confusing.

What do you mean? This is already how you did it.

Reply:
> I am confused as well about why the required values would have default values, because don't think they will ever be used? In that case, for required arguments, supplying the type feels more appropriate.
> 

Read the response I made for @Grazfather 😄 

> Also, for required arguments, shouldn't we use `nargs='+'` instead of `'nargs='*'`. For easy reference (taken from Python docs):
> 
I think `+` shouldn't be a problem, and makes more sense since required. I'll update.

Reply:
> > Regarding what you said about making the required dict in the format , I don't like, I find it ugly and confusing.
> 
> What do you mean? This is already how you did it.

Hum not really, the format in the current implementation is `{"name" : value}` (then `parse_args` does `type(value)` . It's not `{"name" : type }`

Reply:
> Yes, that helps a lot! Do we have any commands that are positional that require a `-` prefix? 

Currently I do not think so.

> I think we could avoid that, and then simply put positional (and variadic) args in the first and optional in the second. Does that work?

I suppose (because I don't see why not 😃 ), but then if I must change the prototype of `parse_arguments` then let's agree for good now about the form it'd take, because I'll have to update each command and test again.

Reply:
> Hum not really, the format in the current implementation is {"name" : value} (then parse_args does type(value) . It's not {"name" : type }

`"name": value` works fine to me.

Reply:
> I suppose (because I don't see why not 😃 ), but then if I must change the prototype of parse_arguments then let's agree for good now about the form it'd take, because I'll have to update each command and test again.

With variadic your extra code makes sense. Let's make it so that positional cannot be flags, except for boolean, basically:

`disasm $pc -hex` makes sense, where hex is a conditional

`asm --arch=x86 nop, nop, nop` makes sense, with variadic positional and one option, but in this case, since each instruction can be more than one 'word' it might make more sense to expect a string, e.g.

`asm --arch=arm "mov r0, 1; mov r1, 0"`

Does this arg parser correct put that whole string into one arg?

Reply:
Yes, that can be done. Then we proscribe the use of variadiac arguments altogether in the argparser right? 
I'm cool with it, just want things to be clear and ok for all of us.

Reply:
I am happy to support variadic

Reply:
We need to make a choice here because we cannot have variadic (as in `nargs='*'`) positional arguments.

Personally I prefer being able to type stuff like `asm --arch=x86 nop; nop; nop` rather than `asm --arch=x86 "nop; nop; nop"` (double quoted) but I can be fine with both. Does someone has a strong (argumented) opinion a way or another?

Reply:
`nop; nop; nop` isn't a great, example. What about
`mov eax, 1; xor ebx, ebx; pop rdi`. Would you go `--arch=x86 "mov eax, 1" "xor ebx, ebx" "pop rdi"`?

If each argument is a word, then how do we get it back into 3 instructions (since we would have EIGHT args, each word)? Just do `" ".join(...)`?.

I am like supporting nargs = `+` or `*`, I just think that when providing asm (another language) it makes sense to provide all of the assembly as one string are, since if arg parse chops it up it'll do one arg per word, and we have to 'rebuild' it into a string anyway.

Basically I think that variadic args are good, but maybe `asm` isn't the best example of where it's handy.

That said, I am still OK with `asm` supporting your way (I just wouldn't use it that way).